File: go/service.payment/internal/payment/service.go
package payment

import (
	"context"
	"fmt"

	"github.com/smiletrl/gateway/service.payment/internal/payment/model"
)

// Service is for main business logic
type Service interface {
	// create a new payment.
	Create(ctx context.Context, req *model.CreatePaymentRequest) error
}

type service struct {
	repo Repository
}

func NewService(repo Repository) Service {
	return &service{repo}
}

func (s *service) Create(ctx context.Context, req *model.CreatePaymentRequest) error {
	// validate request
	if err := req.Validate(); err != nil {
		return fmt.Errorf("error validating: %w", err)
	}

	// create transaction firstly
	transactionID, err := s.repo.CreateTransaction(ctx, req)
	if err != nil {
		return fmt.Errorf("error creating transaction at Create: %w", err)
	}

	// acquirer action
	isApproved := s.isAcquirerApproved(req.Card)

	var newStatus model.TransactionStatus
	if isApproved {
		newStatus = model.TransactionStatusApproved
	} else {
		newStatus = model.TransactionStatusDenied
	}

	// update transaction status as per acquirer's action
	if err := s.repo.UpdateTransactionStatus(ctx, transactionID, newStatus); err != nil {
		return fmt.Errorf("error updating transaction status at Create: %w", err)
	}
	return nil
}

File: python/service.py
from repository import Repository
from model import *

class Service:
    repo: Repository

    def __init__(self, repo: Repository) -> None:
        self.repo = repo

    def create(self, body: Transaction):
        # create transaction firstly
        try:
            tran_id: int = self.repo.createTransaction(body)
        except Exception as e:
            raise Exception("error creating transaction", e)
    
        # if transaction should be approved
        approved: bool = self.isAcquirerApproved(body.card)

        # update transaction status
        try:
            status = TransactionStatus.APPROVED
            if not approved:
                status = TransactionStatus.DENIED
            self.repo.updateTransactionStatus(tran_id, status)
        except Exception as e:
            raise Exception("error updating transaction status")

    # should this transaction be approved
    def isAcquirerApproved(self, card: str) -> bool:
        last_digit = card[-1]
        return int(last_digit) % 2 == 0

